### PR TITLE
[Reviewer: Alex] Sto1465

### PR DESCRIPTION
**Pull request by eleanor-merry**
Alex, can you review this change to add support for using a dictionary in compression/decompression?
The interesting changes are in lz4ruby.c and lz4-ruby.rb

### CODE REVIEW

File: ext/lz4ruby/lz4ruby.c [hunk 1]
```diff
@@ -126,6 +126,90 @@ static VALUE lz4internal_compressHC(VALUE self, VALUE header, VALUE input, VALUE
...
+  const char *header_p;
+  int header_size;
+
+  VALUE result;
+  char *buf;
+  int buf_size;
+
+  int comp_size;
+
+  LZ4_stream_t* const lz4Stream = LZ4_createStream();
```
Comment (AlexHockey) [inline ext/lz4ruby/lz4ruby.c]: It's a bit weird this this is code is so far away from the code that uses the stream.

File: ext/lz4ruby/lz4ruby.c [hunk 2]
```diff
@@ -126,6 +126,90 @@ static VALUE lz4internal_compressHC(VALUE self, VALUE header, VALUE input, VALUE
...
+  LZ4_resetStream(lz4Stream);
```
Comment (AlexHockey) [inline ext/lz4ruby/lz4ruby.c]: Why do we need to reset the stream straight after creating it? Is this a quirk of LZ4?

File: ext/lz4ruby/lz4ruby.c [hunk 3]
```diff
@@ -126,6 +126,90 @@ static VALUE lz4internal_compressHC(VALUE self, VALUE header, VALUE input, VALUE
   return compress_internal(LZ4_compressHC, header, input, in_size);
 }
 
+static VALUE lz4internal_compress_with_dict(VALUE self,
+                                            VALUE header,
+                                            VALUE input,
+                                            VALUE in_size,
+                                            VALUE dictionary,
+                                            VALUE dictSize) {
```
Comment (AlexHockey) [inline ext/lz4ruby/lz4ruby.c]: I don't see why you need to pass the `dictionary` and the `dictSize` when you don't do the same for the `header` (TBH I'm not sure why the library allows you to pass `in_size` as a different parameter but maybe that's for compressing only part of a stream??)

File: ext/lz4ruby/lz4ruby.c [hunk 4]
```diff
@@ -126,6 +126,90 @@ static VALUE lz4internal_compressHC(VALUE self, VALUE header, VALUE input, VALUE
...
+
+  Check_Type(header, T_STRING);
+  header_p = RSTRING_PTR(header);
+  header_size = RSTRING_LEN(header);
+
+  result = rb_str_new(NULL, buf_size + header_size);
+  buf = RSTRING_PTR(result);
+
+  memcpy(buf, header_p, header_size);
+  LZ4_loadDict(lz4Stream, (const char*)dictionary, (int)dictSize);
```
Comment (AlexHockey) [inline ext/lz4ruby/lz4ruby.c]: Hold on... isn't dictionary a _ruby_ string at this point?

Comment (AlexHockey) [inline ext/lz4ruby/lz4ruby.c]: Same comment in the function below.

File: lib/lz4-ruby.rb [hunk 5]
```diff
@@ -5,7 +5,7 @@
 elsif RUBY_PLATFORM == 'java'
   require 'lz4-jruby'
 else
-  require 'lz4ruby'
+  require_relative 'lz4ruby.so'
```
Comment (AlexHockey) [inline lib/lz4-ruby.rb]: Why is this change needed? It suggests the gem isn't installing properly.

File: lib/lz4-ruby.rb [hunk 6]
```diff
@@ -31,17 +31,42 @@ def self.decompress(input, in_size = nil, encoding = nil)
...
+  end
+
   # @deprecated Use {#decompress} and will be removed.
   def self.uncompress(input, in_size = nil)
     return decompress(input, in_size)
   end
 
   private
-  def self._compress(input, in_size, high_compression)
+  def self._compress(input, in_size, high_compression, dict = nil)
```
Comment (AlexHockey) [inline lib/lz4-ruby.rb]: The API to this function is confusing. It implies that `high_compression` and `dict` are orthogonal, but they aren't. Do we need to go through this function. TBH I don't know why it even exists.

**Comment by AlexHockey** [comment]
I've reviewed. My main concern is the way that dictionaries get handled, though I've got a few other comments too.

File: lib/lz4-ruby.rb (hunk 5 above)
Comment (eleanor-merry) [inline lib/lz4-ruby.rb]: I've removed this (it was only needed for how I was testing the changes)
